google/cloud/spanner_v1/merge_result_set.py

```diff
@@ -0,0 +1,114 @@
+from concurrent.futures import ThreadPoolExecutor
+from dataclasses import dataclass
+from queue import Queue
+from typing import Any, TYPE_CHECKING
+from threading import Lock
+
+if TYPE_CHECKING:
+    from google.cloud.spanner_v1.database import BatchSnapshot
+
+QUEUE_SIZE_PER_WORKER = 32
+LOCK = Lock()
+
+
+def _set_metadata(merged_result_set, results):
+    LOCK.acquire()
+    try:
+        merged_result_set._metadata = results.metadata
+        merged_result_set._stats = results.stats
+        merged_result_set._first_row_read = True
+    finally:
+        LOCK.release()
+
+
+class PartitionExecutor:
+    """
+    Executor that executes single partition on a separate thread and inserts
+    rows in the queue
+    """
+
+    def __init__(self, batch_snapshot, partition_id, merged_result_set):
+        self._batch_snapshot: BatchSnapshot = batch_snapshot
+        self._partition_id = partition_id
+        self._merged_result_set: MergedResultSet = merged_result_set
+        self._queue: Queue[PartitionExecutorResult] = merged_result_set._queue
+
+    def run(self):
+        try:
+            results = self._batch_snapshot.process_query_batch(self._partition_id)
+            merged_result_set = self._merged_result_set
+            for row in results:
+                if not merged_result_set._first_row_read:
+                    _set_metadata(merged_result_set, results)
+                self._queue.put(PartitionExecutorResult(data=row))
+            # Special case: The result set did not return any rows.
+            # Push the metadata to the merged result set.
+            if not merged_result_set._first_row_read:
+                _set_metadata(merged_result_set, results)
+        except Exception as ex:
+            self._queue.put(PartitionExecutorResult(exception=ex))
+        finally:
+            self._queue.put(PartitionExecutorResult(is_last=True))
+
+
+@dataclass
+class PartitionExecutorResult:
+    data: Any = None
+    exception: Exception = None
+    is_last: bool = False
+
+
+class MergedResultSet:
+    """
+    Executes multiple partitions on different threads and then combines the
+    results from multiple queries using a synchronized queue. The order of the
+    records in the MergedResultSet is not guaranteed.
+    """
+
+    def __init__(self, batch_snapshot, partition_ids, max_parallelism):
+        self._metadata = None
+        self._stats = None
+        partition_ids_count = len(partition_ids)
+        self._finished_counter = partition_ids_count
+        # True if at least one row has been read from any of the partition
+        self._first_row_read = False
+        if max_parallelism != 0:
+            parallelism = min(partition_ids_count, max_parallelism)
+        else:
+            parallelism = partition_ids_count
```

Comment:
In python only 1 cpu core is used and as its an I/O bound operation so IMO cpu cores should not matter. Should we limit the `parallelism` to some max number (for ex 1000)?

Reply:
I think that we should limit this, because this unlimited version means that it will try to execute for example 1,000 queries in parallel if there are 1,000 partitions. That will eat a lot of memory.

The query execution is an I/O operation, meaning that even though only one core is used, all the 1,000 query executions could be started.

I think that we should put a much smaller hard limit than 1,000. I feel that the default parallelism value used above (32) is a more reasonable hard limit, unless we can get our hands on some dynamic value that tells us that the current host is probably able to do more than than.

Reply:
32 (QUEUE_SIZE_PER_WORKER) is not the parallelism but the max number of messages per thread that can be inserted. 
Adding a new variable MAX_PARALLELISM with value as 100. Please let me know if tis sounds good